consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/blockcreation/backward/sync/BackwardSyncStep.java

```diff
@@ -0,0 +1,137 @@
+    return future
+        .thenApply(this::earliestUnprocessedHash)
+        .thenCompose(this::requestHeader)
+        .thenApply(this::saveHeader)
+        .thenApply(this::possibleMerge)
+        .thenCompose(this::possiblyMoreBackwardSteps);
+  }
+
+  @VisibleForTesting
+  protected Hash earliestUnprocessedHash(final Void unused) {
+    BackwardChain backwardChain = context.getBackwardChainMap().get(pivot.getHeader().getNumber());
+    Hash parentHash = backwardChain.getFirstHeader().getParentHash();
+    LOG.debug(
+        "First unprocessed hash for current pivot is {} expected on height {}",
+        () -> parentHash.toString().substring(0, 20),
+        () -> backwardChain.getFirstHeader().getNumber() - 1);
+    return parentHash;
+  }
+
+  @VisibleForTesting
+  protected CompletableFuture<BlockHeader> requestHeader(final Hash hash) {
+    LOG.debug("Requesting header for hash {}", () -> hash.toString().substring(0, 20));
+    return GetHeadersFromPeerByHashTask.forSingleHash(
+            context.getProtocolSchedule(),
+            context.getEthContext(),
+            pivot.getHeader().getHash(),
+            pivot.getHeader().getNumber(),
+            context.getMetricsSystem())
+        .run()
+        .thenApply(
+            peerResult -> {
+              BlockHeader blockHeader = peerResult.getResult().get(0);
+              LOG.debug(
+                  "Got header {} with height {}",
+                  () -> blockHeader.getHash().toString().substring(0, 20),
+                  blockHeader::getNumber);
+              return blockHeader;
+            });
+  }
+
+  @VisibleForTesting
+  protected BlockHeader saveHeader(final BlockHeader blockHeader) {
+    BackwardChain backwardChain = context.getBackwardChainMap().get(pivot.getHeader().getNumber());
+    backwardChain.saveHeader(blockHeader);
+    return blockHeader;
+  }
+
+  @VisibleForTesting
+  protected BlockHeader possibleMerge(final BlockHeader blockHeader) {
+    BackwardChain backwardChain = context.getBackwardChainMap().get(pivot.getHeader().getNumber());
+    Optional<BackwardChain> maybeHistoricalBackwardChain =
+        Optional.of(context.getBackwardChainMap().get(backwardChain.getFirstHeader().getNumber()));
+    maybeHistoricalBackwardChain.ifPresent(backwardChain::merge);
+    return backwardChain.getFirstHeader();
+  }
+
+  // if the previous header is not present yet, we need to go deeper
+  @VisibleForTesting
+  protected CompletableFuture<Void> possiblyMoreBackwardSteps(final BlockHeader blockHeader) {
+    CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+    if (context.getProtocolContext().getBlockchain().contains(blockHeader.getHash())) {
+      LOG.debug(
+          "The backward sync let us to a block that we already know... We will init forward sync...");
+      completableFuture.complete(null); // we finished backward sync
+      return completableFuture;
```

Comment:
this is an interesting choice - to hit storage looking for the blockhash rather than comparing it to the head or the latest finalized head.  I think this could be a problem if we got a malicious request to executeBlock.  We could end up on a big reorg fork.

Ideally we would check that the block parent is either our current head or latest finalized block.  

Also we should throw rather than more backwards steps if 
* If we DO have a finalized block, and the current block number is not less than finalized block number
* If we DO NOT have a finalized block, keep recursing until we are within some threshold of the terminal total difficulty block.  

Re: the second bullet, I think we might have a corner case where we reorg back before the terminal total difficulty block, so we need to have some reasonable provision for this, but not try to backwards sync all the way back to genesis.  We should abort a backward sync that exceeds that threshold

Reply:
Also, same applies to the head.  It will be normal to need to sync back to a block with a parent prior to the current head.  Anything that is after the latest finalized block is fair game for a reorg